### ISSUE TITLE
Machine Learning Interview-Decision Tree

### ISSUE THREAD

**Issue by Eamymao** [Eamymao/winter-ann-arbor#4]
### Idea:
Each sample will finally fall into one leaf node of the tree. Then, each leaf node will have its own conditional probability:
Given the observation X, the probability that the sample belongs to class k: $P(Y=k|X)$

The conditional probability usually tend to tilt towards to one class $k$, that is there is higher probability that samples in that leaf node belong to that class  $k$. When we use decision tree to classify samples, the model will classify the new observation which falls into that node to be class  $k$

### Loss Function
The maximum likelihood function with regularization. Our training goal is to minimize loss function.

### Learning Method:
At each node of a tree, we recursively select an optimal feature to split our training data falling in current node into subsets. If the subset has almost been correctly classified, we initialize a new leaf node to store the subset. If not, we will choose new optimal feature for splitting until all subsets are almost correctly classified or there is no feature to select for split. 

However, to improve models' generalization ability we need to prune the tree from bottom to top. That is, eliminate too small leaf node and return them to the higher node.

In the end, each decision tree represent a distribution of probability.

### Feature selection
Feature selection is the process to choose one selection to split feature space, which is based on information gain or information gain ratio.
Information gain is a metric to evaluate how good is the subset classified.

#### Information gain
Entropy is a metric to evaluate uncertainty of random variable. Suppose that $X$ has limited values $x_i$ whose probability distribution is 
Then, entropy of $X$ is

$$H(X) = -\sum_{i=1}^np_ilog(p_i)$$
 which only relies on distribution of $X$ rather than its value.
 
$$H(Y|X)$$represents given value of $X$, the uncertainty of $Y$.
$$H(Y|X)=\sum_{i=1}^np_iH(X=x_i)$$ where $p_i=P(X=x_i), i =1,2...n$.(X给定条件下Y的条件概率分布的熵的数学期望)。

Information gain of feature A for training data D, denoted as $g(D,A) = H(D)-H(D|A)$. This is called mutual information. It represents the decreased uncertainty of classification of data D because of feature A. 决策树学习中的信息增益等价于训练数据集中类与特征的互信息。信息增益代表由于特征A二十的对数据集D的分类的不确定性的减少程度。

根据信息增益准则的特征选择方法是：对训练集/子集D,计算其每个特征的信息增益，并比较他们的大小，选择信息增益最大的特征。

#### Computation of mutual information
Suppose training data set is $D$,$|D|$represent sample size of training set. There are K classes, each of which is denoted as $C_k$,$|C_k|$denotes the size of class k.
Suppose feature A has n different values,${a_1,a_2,...,a_n}$.According to feature A, we divide D into n subsets, $D_1,D_2,...D_n$. Let's denote set of samples which belongs to class $C_k$ in subset$D_i$, $D_{ik}$.

$$H(D) = -\sum_{k=1}^K\frac{|C_k|}{|D|}log_2\frac{|C_k|}{|D|}$$,

$$H(D|A)=\sum_{i=1}^n\frac{|D_i|}{|D|}H(|D_i|)=-\sum_{i=1}^n\frac{|D_i|}{|D|}\sum_{k=1}^K\frac{|C_k|}{|D_{ik}|}log_2\frac{|C_k|}{|D_{ik}|}$$

$$g(D,A)=H(D)-H(D|A)$$

**Information gain ratio** is another metric to evaluate which feature should be selected,
$$g_R(D|A)=\dfrac{g(D,A)}{H(D)}$$

### Tree generation algorithm
每次分裂都计算各特征对训练集D的信息增益比，选择信息增益比最大的特征$A_g$;如果$A_g$的信息增益比小于阈值$\epsilon$,则置T为单节点树，并将D中实例树最大的类$C_k$作为该节点的类，返回T；
否则，对$A_g$的每一可能值$a_i$,依照$A_g=a_i$分割成子集，即若干个非空$D_i$，将$D_i$中实例最多的类作为标记，构建子节点。然后对新的子节点以$D_i$为训练集，以$A-A_g$作为特征集，再循环前面的内容。

### Pruning Tree
决策树的剪枝是通过极小化决策树整体的损失函数来实现。
$$C_\alpha(T)=C(T)+\alpha|T|$$，其中$C_(T)$表示模型与训练数据的拟合程度，即预测误差，|T|表示模型复杂度。剪枝则是在惩罚项系数固定的情况下，选择损失函数最小的子树。所以决策树剪枝通过优化损失函数还考虑了减小模型复杂度的方面。

### Cart算法
